src/backend/distributed/transaction/distributed_transaction_management.c

```diff
@@ -0,0 +1,431 @@
+/*-------------------------------------------------------------------------
+ *
+ * distributed_transaction_management.c
+ *
+ *  Infrastructure for distributed transaction management and distributed deadlock
+ *  detection.
+ *
+ * Copyright (c) 2017, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "miscadmin.h"
+
+#include "funcapi.h"
+#include "access/htup_details.h"
+#include "catalog/pg_type.h"
+#include "datatype/timestamp.h"
+#include "distributed/distributed_transaction_management.h"
+#include "distributed/metadata_cache.h"
+#include "nodes/execnodes.h"
+#include "storage/ipc.h"
+#include "storage/lwlock.h"
+#include "storage/proc.h"
+#include "storage/spin.h"
+#include "storage/s_lock.h"
+#include "utils/timestamp.h"
+
+
+/*
+ * Each backend's active distributed transaction data reside in the
+ * shared memory on the DistributedTransactionShmemData.
+ */
+typedef struct DistributedTransactionShmemData
+{
+	int trancheId;
+#if (PG_VERSION_NUM >= 100000)
+	NamedLWLockTranche namedLockTranche;
+#else
+	LWLockTranche lockTranche;
+#endif
+	LWLock lock;
+
+	/*
+	 * We prefer to use an atomic integer over sequences for two
+	 * reasons (i) orders of magnitude performance difference
+	 * (ii) allowing read-only replicas to be able to generate ids
+	 */
+	pg_atomic_uint64 nextTransactionId;
+
+	DistributedTransactionBackendData sessions[FLEXIBLE_ARRAY_MEMBER];
+} DistributedTransactionShmemData;
+
+
+static shmem_startup_hook_type prev_shmem_startup_hook = NULL;
+static DistributedTransactionShmemData *distributedTransactionShmemData = NULL;
+static DistributedTransactionBackendData *MyDistributedTransactionBackend = NULL;
+
+
+static void DistributedTransactionManagementShmemInit(void);
+static size_t DistributedTransactionManagementShmemSize(void);
+
+
+PG_FUNCTION_INFO_V1(assign_distributed_transaction_id);
+PG_FUNCTION_INFO_V1(get_distributed_transaction_id);
+
+
+/*
+ * assign_distributed_transaction_id updates the shared memory allocated for this backend
+ * and sets initiatorNodeIdentifier, transactionId, timestamp fields with the given
+ * inputs. Also, the function sets the database id via MyDatabaseId variable provided
+ * by "misadmmin.h".
+ */
+Datum
+assign_distributed_transaction_id(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	/* MyDistributedTransactionBackend should always be avaliable, just out of paranoia */
+	if (!MyDistributedTransactionBackend)
+	{
+		ereport(ERROR, (errmsg("backend is not ready for distributed transactions")));
+	}
+
+	SpinLockAcquire(&MyDistributedTransactionBackend->mutex);
+
+	MyDistributedTransactionBackend->databaseId = MyDatabaseId;
```

Comment:
What I'm thinking is that we don't want to allow a distributed tx to be assigned when one is already assigned. What I'm trying to say above is that you can do the "already assigned and error if so" before acquiring the spinlock, because nobody else could write to it.